File: dashlane-project/Dashlane/src/main/java/com/dashlane/DashlaneApplication.java
package com.dashlane;

import android.app.ActivityManager;
import android.app.Application;
import android.content.Context;
import android.os.Build;
import android.os.Looper;
import android.webkit.WebView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.hilt.work.HiltWorkerFactory;
import androidx.work.Configuration;

import com.braze.ui.inappmessage.BrazeInAppMessageManager;
import com.dashlane.debug.DeveloperUtilities;
import com.dashlane.hermes.LogRepository;
import com.dashlane.util.BuildContract;

import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.HiltAndroidApp;

@HiltAndroidApp
public class DashlaneApplication extends Application implements Configuration.Provider {

    static {
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
    }

    @Inject
    ApplicationObserver dashlaneObserver;

    @Inject
    LogRepository logRepository;

    @Inject
    HiltWorkerFactory workerFactory;

    @Override
    public void onCreate() {
        super.onCreate();

        BrazeInAppMessageManager.getInstance().ensureSubscribedToInAppMessageEvents(getApplicationContext());

        updateWebViewDataDirectory();
        if (isMainApplicationThread()) {
            dashlaneObserver.onCreate(this);
        }
    }

    @Override
    public void onTerminate() {
        if (isMainApplicationThread()) {
            dashlaneObserver.onTerminate(this);
        }
        super.onTerminate();
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        
        
        if (!DeveloperUtilities.isUnitTestMode()) {
            BuildContract.initializeSystemSettings();
        }
    }

    private void updateWebViewDataDirectory() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
            
            return;
        }
        String processName = getProcessName();
        if (processName == null || "com.dashlane".equals(processName)) {
            
            return;
        }
        
        
        WebView.setDataDirectorySuffix("webview-" + processName);
    }

    private boolean isMainApplicationThread() {
        Looper mainLooper = Looper.getMainLooper();
        return mainLooper != null && mainLooper.isCurrentThread() && "com.dashlane".equals(getMyProcessName());
    }

    private String getMyProcessName() {
        int pid = android.os.Process.myPid();
        ActivityManager manager = (ActivityManager) this.getSystemService(Context.ACTIVITY_SERVICE);
        if (manager == null) return null;
        List<ActivityManager.RunningAppProcessInfo> processInfoList = manager.getRunningAppProcesses();
        if (processInfoList == null) return null;
        for (ActivityManager.RunningAppProcessInfo processInfo : processInfoList) {
            if (processInfo.pid == pid) return processInfo.processName;
        }
        return null;
    }

    @NonNull
    @Override
    public Configuration getWorkManagerConfiguration() {
        return new Configuration.Builder()
            .setMinimumLoggingLevel(android.util.Log.INFO)
            .setWorkerFactory(workerFactory)
            .build();
    }
}

File: dashlane-project/Dashlane/src/main/java/com/dashlane/DashlaneApplication.kt
package com.dashlane

import android.app.ActivityManager
import android.app.Application
import android.os.Looper
import android.os.Process
import android.util.Log
import android.webkit.WebView
import androidx.appcompat.app.AppCompatDelegate
import androidx.hilt.work.HiltWorkerFactory
import androidx.work.Configuration
import com.braze.ui.inappmessage.BrazeInAppMessageManager
import com.dashlane.debug.services.DaDaDaBase
import dagger.hilt.android.HiltAndroidApp
import javax.inject.Inject

@HiltAndroidApp
class DashlaneApplication : Application(), Configuration.Provider {
    @Inject
    lateinit var dashlaneObserver: ApplicationObserver

    @Inject
    lateinit var workerFactory: HiltWorkerFactory

    @Inject
    lateinit var dadada: DaDaDaBase

    override val workManagerConfiguration: Configuration
        get() = Configuration.Builder()
            .setMinimumLoggingLevel(Log.INFO)
            .setWorkerFactory(workerFactory)
            .build()

    private val isMainApplicationThread: Boolean
        get() {
            val mainLooper = Looper.getMainLooper()
            return mainLooper != null && mainLooper.isCurrentThread && "com.dashlane" == myProcessName
        }

    private val myProcessName: String?
        get() {
            val pid = Process.myPid()
            val processInfoList = (this.getSystemService(ACTIVITY_SERVICE) as? ActivityManager)
                ?.runningAppProcesses
                ?: return null
            for (processInfo in processInfoList) {
                if (processInfo.pid == pid) return processInfo.processName
            }
            return null
        }

    init {
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
    }

    override fun onCreate() {
        super.onCreate()

        BrazeInAppMessageManager.getInstance().ensureSubscribedToInAppMessageEvents(applicationContext)
        dadada.refreshAsync(this)

        updateWebViewDataDirectory()
        if (isMainApplicationThread) {
            dashlaneObserver.onCreate(this)
        }
    }

    override fun onTerminate() {
        if (isMainApplicationThread) {
            dashlaneObserver.onTerminate(this)
        }
        super.onTerminate()
    }

    private fun updateWebViewDataDirectory() {
        val processName = getProcessName()
        if (processName == null || "com.dashlane" == processName) {
            
            return
        }
        
        
        WebView.setDataDirectorySuffix("webview-$processName")
    }
}